components/style/matching.rs

```diff
@@ -475,6 +478,20 @@ trait PrivateMatchMethods: TElement {
                 .cancel_transitions_with_nontransitioning_properties(&transitioning_properties);
         }
 
+        {
+            let new_values = Arc::make_mut(new_values);
```

Comment:
Same, it seems this will disable all sorts of style sharing optimizations.

Reply:
Same here. I originally went with an `is_empty()` check, but ended up doing this in the methods themselves to enable more fine-grained control over when the style is cloned.